<a href="https://colab.research.google.com/github/wo1ever/programmers/blob/master/124%20%EB%82%98%EB%9D%BC%EC%9D%98%20%EC%88%AB%EC%9E%90.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 124 나라의 숫자

124 나라가 있습니다. 124 나라에서는 10진법이 아닌 다음과 같은 자신들만의 규칙으로 수를 표현합니다.


- 124 나라에는 자연수만 존재합니다.
- 124 나라에는 모든 수를 표현할 때 1, 2, 4만 사용합니다.
- 예를 들어서 124 나라에서 사용하는 숫자는 다음과 같이 변환됩니다.

| 10진수 | 124진수 |
| ------ | ------- |
| 1      | 1       |
| 2      | 2       |
| 3      | 4       |
| 4      | 11      |
| 5      | 12      |
| 6      | 14      |
| 7      | 21      |
| 8      | 22      |
| 9      | 24      |
| 10     | 41      |


자연수 n이 매개변수로 주어질 때, n을 124 나라에서 사용하는 숫자로 바꾼 값을 return 하도록 solution 함수를 완성해 주세요.

### 제한사항

- n은 500,000,000이하의 자연수 입니다.

### 입출력 예

| n    | result |
| ---- | ------ |
| 1    | 1      |
| 2    | 2      |
| 3    | 4      |
| 4    | 11     |

# 의식의 흐름

처음에는 3진수로 바꾸는 방법과 똑같을 것이라고 생각해서 금방 끝날 줄 알았다.
그런데 0이 아닌 4가 들어가있는 바람에, 단순히 n % 3 이  인덱스가 되는 형식이 아니었다.

숫자를 종이에 나열해 보니, 규칙이 있긴 있었는데 이걸 어떻게 해결해야할까.. 엄청 고민이 되었다.

처음에는 아주아주 직관적으로 생각해보고, 수업때 TDD 해본 것 처럼(ㅋㅋㅋ) 예시를 하나 잡아, ~~통과시켜 보려고 했으나 시킬 수 없었고~~ 대략적인 생각을 해 보았다...

뭔가 규칙을 잘못 찾았기 때문이었다...!

In [0]:
def solution(n):
  N = "124"
  
  num_list = []
  
  div1, mod1 = divmod(n,3)
  
  if mod1 == 1:
    num_list.append(N[0])
  elif mod1 == 2:
    num_list.append(N[1])
  else:
    num_list.append(N[2])

  div2, mod2 = divmod(div1,3)
  if mod2 == 1:
    num_list.append(N[0])
  elif mod2 == 2:
    num_list.append(N[1])
  else:
    num_list.append(N[2])
  
  div3, mod3 = divmod(div2,3)
  if mod3 == 1:
    num_list.append(N[0])
  elif mod3 == 2:
    num_list.append(N[1])
  else:
    num_list.append(N[2])
  
  num_list.reverse()
  
  return ''.join(num_list)

In [97]:
solution(12)

'114'

내 생각처럼 간단하게 풀리는 문제는 아니었던 것이었다..

일단 파이썬의 아주 세세한 함수, 문법 등을 많이 숙지하지 못해서인지 if-else구문밖에 생각이 안났다.

그리고 생각보다 다음 수를 구성하는 규칙이 너무 헷갈렸다..

In [0]:
def solution(n):
  N = "124"  
  num_list = []
  div, mod = divmod(n,3)

  if div == 0:
    if mod == 1:
      num_list.append(N[0])
    elif mod == 2:
      num_list.append(N[1])
    else:
      num_list.append(N[2])
      
  else:
    while div > 0:
      div, mod = divmod(div,3)
  
  num_list.reverse()
  
  return ''.join(num_list)

In [140]:
solution(3)

''

어디에서 어떤 변수가 변경되는 지 찾아내려고 def 함수를 지우고 단계별로 변수가 어떻게 바뀌는 지 찍어 보았다.


> While문은 조건식이 참인 동안만 반복문을 수행하는 루프이다.
조건문이 거짓이 된다면, 루프를 빠져나간다.

이것에 착안하여 여러 시도를 해보았는데, 이것 또한 쉽지 않았다😅

그리고 나서 A4용지를 9쪽이나 썼다.

In [0]:
N = "124"  
num_list = []

div, mod = divmod(15,3)

while True:
  
  if div == 0:
    if mod == 1:
      num_list.append(N[0])
    elif mod == 2:
      num_list.append(N[1])
    else:
      num_list.append(N[2])
    break
    
  else:
    div, mod = divmod(div,3)

num_list

In [169]:
num_list

['1']

코드들을 좀 예쁘게 다듬으려고 N의 순서를 바꿨다.

124 보다는 412로 했을 경우에, mod가 1, 2일 때 --> N[mod] 가 바로 숫자 1, 2를 가리켜서 좀 자연스럽다고 생각되었고 애초에 0을 대체한 수가 4였기 때문에 012 의 순서가 412가 되어야 한다고 생각했다.

그랬더니 아주 조금 나아졌긴 했지만,...

그 뒤로도 오랫동안 고민한 결과,  ~~컨닝~~ 레퍼런스를 참고하기로 했다. 다행히도 역시 나만 어려운게 아니었나보다😭

역시 가장 큰 고민은 mod==0 인 경우를 어떻게 처리할 지 였다. 그치만 아무리해도, 끝자리 수 밖에 못맞춘다.... 왜일까 정말 모르겠다!



내가 구현하고 싶었던 것:

1. n의 div, mod를 구한다.

2. mod가 0이라면, div를 1 낮춘다.

3. n에 div를 넣는다.

4. div가 0이 될 때 까지 1~3을 반복한다.

| n      | div     | mod   | 124진수변환 | 결과 |
| ------ | ------- | ----- | ----------- | ------- |
| **10** | 3       | 1     | ☞ 1         |         |
| 3      | 1       | **0** | ☞ 4         |         |
|        |→ 0 |       |             |         |
| 0 (X)  |         |       |             | **41**  |
|        |         |       |             |         |
| **13** | 4       | 1     | ☞ 1         |         |
| 4      | 1       | 1     | ☞ 1         |         |
| 1      | 0       | 1     | ☞ 1         |         |
| 0 (X)  |         |       |             | **111** |
|        |         |       |             |         |
| **15** | 5       | **0** | ☞ 4         |         |
|        |→ 4 |       |             |         |
| 4      | 1       | 1     | ☞ 1         |         |
| 1      | 0       | 1     | ☞ 1         |         |
| 0 (X)  |         |       |             | **114** |
|        |         |       |             |         |
| **16** | 5       | 1     | ☞ 1         |         |
| 5      | 1       | 2     | ☞ 2         |         |
| 1      | 0       | 1     | ☞ 1         |         |
| 0 (X)  |         |       |             | **121** |
|        |         |       |             |         |
|        |         |       |             |         |



In [0]:
def solution(n):
  N = "412"
  num_list = []
  div, mod = divmod(n,3)
  
  while div > 0:

    if mod == 0:
      div -= 1
      num_list.append(N[mod])
    else:
      num_list.append(N[mod])

      div, mod = divmod(div,3)
  
  num_list.reverse()
  return ''.join(num_list)

In [202]:
# test
print(solution(1))
print(solution(2))
print(solution(3))
print(solution(4))
print(solution(5))

print(solution(8))
print(solution(11))
print(solution(13))
print(solution(14))



4
1
2
2
42
11
12


나름의 test코드까지 짜면서 열심히 해 보았지만, 내 코드는 왜 끝자리밖에 안나올까... 슬프게도 조금 더 레퍼런스를 찾아보았다. 검색하다보니, 내가 정말 머릿속 어딘가에서 내가 구현하고 싶었던 코드를 아주 간단하게 짜놓으신 것을 보았다. GG....(= God Google)

[감사합니다](http://oneshottenkill.tistory.com/128)

?!.....!!!!!!!

역시.. 이 분야는 정말 배울게 너무 많다...ㅠㅠ 이렇게 배우다가는 인생 80년 살아도 모자랄 것 같다는 생각이 들었다.

그래도 어떤 식으로 내가 고민했고, 점진적으로 코드를 고쳐나가면서, 알고리즘을 수정하고 많은 연습이 되었던 것 같다.

앞으로 더 많이 풀어보고 연습하면 나아~~질 테니까!~~ 졌으면 좋겠다.

In [204]:
def change124(n):
  
    answer = ""
    
    i=0

    while n>0:

        i = n%3

        n //= 3

        if i==0:

            n-=1

        answer = '412'[i] + answer

    return answer



# 아래는 테스트로 출력해 보기 위한 코드입니다.

print(change124(10))

# 출처: http://oneshottenkill.tistory.com/128 [copyndpaste .cf]

41
